### ISSUE TITLE
Value-diff compression

### ISSUE THREAD

**Issue by UkoeHB** [projectharmonia/bevy_replicon#178]
Some networked games try to compress diffs between updates by networking only the value difference at a component level. Replicon doesn't currently support that.

Here is a first-draft idea for how to support it:

If we pass the change-limit tick + current tick into serializers, then the serializer can have a local that tracks historical state and performs a diff internally. The deserializer would also need to track historical state and apply diffs to the correct older value. We might be able to provide a pre-baked value-diff compressor for types that implement certain traits.

Note that supporting this would require changes to the shared copy buffer, since different clients could get different component serializations.

**Comment by Shatur** [comment]
### Current state

We now pass ticks to serializers and send different data to different clients. We just need to add a way to get an old value.

### Component-based approach

We can create a component `ServerHistory`. It's not generic itself, but provides `get_mut::<D>()`/`insert(D)`.
`D` may or may not be equal to the component we serialize (to let user choose how to store the value).
Internally it stores a hashmap with `ClientId` and blob data that does type-erasure.
We add `ClientId` and `&mut ServerHistory` fields to  `SerializeCtx` for the access.
 
The overhead of this solution will be an add additional lookup for `ServerHistory` per-entity + iteration on all components on disconnect to cleanup. Getting history will require additional lookup for `D`.

 I also considered making it generic to avoid dealing with type-magic, but it makes it harder to clean when a client disconnects, requires to query for `ServerHistory` for each component instead of once per entity and I can't use it with `SerializeCtx` this way.

### Resource-based approach

Exactly the same as component-based approach, but with additional internal hash map for entities, no `Option` (the resource will be always present) and `Entity` field inside `SerializeCtx` (to make lookups).

The iteration overhead will be near-zero (we lookup for the resource once instead of per-entity), but if a component serialization function needs the history, it's a bit slower due to the entity-based lookup. Also simpler cleanup on disconnect.

### Questions

I would appreciate opinions.

* Component-based or resource-based approach? Originally I thought about component-based one because it's per-entity. But after diving deeper into performance considerations, I think resource-based approach will be better. Most of the time users won't need access to the history, so we probably want to optimize for it.
* How to erase types? Bevy exposes [Resources](https://docs.rs/bevy/latest/bevy/ecs/storage/struct.Resources.html), but doesn't provide a way to insert a new resource. We can implement our own abstraction by using [SparseSet](https://docs.rs/bevy/latest/bevy/ecs/storage/struct.SparseSet.html) with [ResourceData](https://docs.rs/bevy/latest/bevy/ecs/storage/struct.ResourceData.html), but we also don't need ticks. So maybe implement our own abstraction, similar to `BlobVec` in Bevy (not exposed to users)?

**Comment by Shatur** [comment]
> Can't we just map to ComponentId?

Yeah, that's what I suggesting: map `ComponentId` into our implementation of `BlobVec` (I don't think we can just map it to `OwningPtr`).

**Comment by Shatur** [comment]
### Additional notes

Instead of providing the resource from context, we provide a resource adapter that is scoped to the current entity. Just to avoid passing entity each time, will be nicer to use.

Forgot to talk about ticks. We will need to store values with ticks and return only the last acked. So instead of `get_mut::<D>()`/`insert(D)`, we probably want `last_acked::<D>()`/`add_unacked(D, tick)`. Internally we will cleanup acked values.

### API showcase

Considering all the above, here is how the API might look like:

```rust
#[derive(Component, Serialize, Deserialize, Clone, Copy)]
struct Example {
    a: u32,
    b: u32,
    c: u32,
}

// Let's write the diff logic as methods for clarity.
impl Example {
    fn delta(&self, _other: &Self) -> ExampleDelta {
        todo!();
    }

    fn apply(&mut self, _delta: ExampleDelta) {
        todo!();
    }
}

/// Depends on how user want to compress, but let's consider the simplest example.
#[derive(Serialize, Deserialize, Clone, Copy)]
struct ExampleDelta {
    a: Option<u32>,
    b: Option<u32>,
    c: Option<u32>,
}

/// User overrides the serialization function that checks last received value using context.
fn serialize(
    ctx: &SerializeCtx,
    component: &Example,
    cursor: &mut Cursor<Vec<u8>>,
) -> bincode::Result<()> {
    if let Some(acked_component) = ctx.component_values.last_acked::<Example>() {
        let delta = component.delta(acked_component);
        DefaultOptions::new().serialize_into(cursor, &delta)?;
    } else {
        // We can call the default serialization function when there are no delta.
        default_serialize(ctx, component, cursor)?;
    }

    ctx.component_values.add_unacked(*component, ctx.server_tick);

    Ok(())
}

/// Deserialization in place called only for already received values, so we override it to use delta.
///
/// Regular deserialization function in this example will be default one.
fn deserialize_in_place(
    _deserialize: DeserializeFn<Example>,
    _ctx: &mut WriteCtx,
    component: &mut Example,
    cursor: &mut Cursor<&[u8]>,
) -> bincode::Result<()> {
    let delta: ExampleDelta = DefaultOptions::new().deserialize_from(cursor)?;
    component.apply(delta);
    Ok(())
}
```

**Comment by malfuu** [comment]
Wanted to chime in, as I have 4KiB components that could use value-diffing, as they stretch bandwidth to 800KiB/s on a 30 UPS server. 

Firstly, I do have a question: do you pretend to only send value-diffs throughout a component's lifetime? What about intermittently sending snapshots at user-defined tick intervals?

I was thinking of a simpler component-based approach, where history is cached:
```rust
#[derive(Component)]
struct Cached<T> {
    cached: T,
    last_tick: RepliconTick
}
```

where as a user can register like (how diff is done is TODO):
```rust
app.replicate_diff::<Foo>(30); // send a new snapshot every 30 UPS
```

Where diffs are calculated off the cached value, where it is refreshed ever so often.

I'm not expert in networking or type erasure techniques, especially as I struggle knowing there are many diff strategies.

**Comment by Shatur** [comment]
> Firstly, I do have a question: do you pretend to only send value-diffs throughout a component's lifetime? What about intermittently sending snapshots at user-defined tick intervals?

Not sure If I get your questions 🤔 

But I want to clarify that each client should have its own snapshot. That's because a mutation could have dropped on the wire and needs to be resend. So one client might receive a mutation for tick X, while others might not. This is one of the reasons why it's quite tricky to get right...

If you have large components - I'd highly recommend to split the data into smaller components.
Alternatively - you can add compression per-component by customizing the serialization.

**Comment by malfuu** [comment]
What I meant by sending value-diffs through a component's lifetime is if you only send one snapshot when the object is first replicated. I was instead thinking of sending a new snapshot ever so often, analogous to video streaming.

**Comment by Shatur** [comment]
Ah, you're suggesting sending diffs relative to some value in time and periodically updating this value?

**Comment by malfuu** [comment]
Exactly so, but I was forgetting that the server tries to resend component updates in case they are dropped.
I would like to see if losing some updates in this replication method would be acceptable.

**Comment by Shatur** [comment]
I think this will be even trickier to implement 😅 It's easier to just compare to the last acked value.

**Comment by Shatur** [comment]
However, this requires tracking this last acked value for each client 🤔 
With what you're suggesting we'll be able to track it globally (one value for all clients). And update it once all clients actually received the value. I think I like this idea.

**Comment by Shatur** [comment]
I.e. we check if client received the update. If yes - good.
If no, instead of sending the whole component - diff it with the last known value all clients received.

**Comment by malfuu** [comment]
> And update it once all clients actually received the value.

I'm not sure what you mean by all updating when all clients received the value. Does that mean we only send diffs when all clients have received it? What about clients for which the entity is not visible?

Anyways, I could create a quick draft PR with the API (excluding implementation) and some tests, and after 0.19 we could look further into it, if it is interesting to you.

**Comment by Shatur** [comment]
> I'm not sure what you mean by all updating when all clients received the value. Does that mean we only send diffs when all clients have received it?

Right, under "all clients" I mean "all clients that can see the component/entity".
Just to clarify: you suggested to not track a diff for each client separately and instead have only a single value for which the diff is computed. How, how to select which value to use for this diff? And I suggested to use the oldest value that is confirmed by all clients that saw it.

> Anyways, I could create a quick draft PR with the API (excluding implementation) and some tests, and after 0.19 we could look further into it, if it is interesting to you.

Sure. Yeah, I'd prefer to have changes like this after 0.19 release.

**Comment by malfuu** [comment]
Is there any diffing mechanism that could be applied for this?
There are crates such as `daft` or `difficient`, but I think a simple trait provided by replicon is enough -- if we provide a trait, that is.

**Comment by Shatur** [comment]
I don't mind using a crate to improve user experience 🤔 But it would be best to check what is already in the dependency tree in Bevy.
